src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -89,12 +90,201 @@ struct extent_to_write_t {
+    :type(type_t::DATA), addr(addr), len(len), bl(bl) {}
+
+  extent_to_insert_t(laddr_t addr, extent_len_t len)
+    :type(type_t::ZERO), addr(addr), len(len) {}
+};
+using extent_to_insert_list_t = std::list<extent_to_insert_t>;
+
+// Encapsulates extents to be retired in do_removals.
+using extent_to_remove_list_t = std::list<LBAMappingRef>;
+
+struct overwrite_ops_t {
+  extent_to_remap_list_t to_remap;
+  extent_to_insert_list_t to_insert;
+  extent_to_remove_list_t to_remove;
+};
+
+// prepare to_remap, to_retire, to_insert list
+overwrite_ops_t prepare_ops_list(
+  lba_pin_list_t &pins_to_remove,
+  extent_to_write_list_t &to_write) {
+  assert(to_write.size() != 0 && pins_to_remove.size() != 0);
+  overwrite_ops_t ops;
+  ops.to_remove.swap(pins_to_remove);
+  auto& front = to_write.front();
+  auto& back = to_write.back();
+
+  // overwrite happens in one original extent.
+  if (ops.to_remove.size() == 1 &&
+    (front.is_existing() || back.is_existing())) {
+    if (front.is_existing()) {
+      assert(to_write.size() == 2 || to_write.size() == 3);
+      assert(to_write.size() == 3 ? back.is_existing() : true);
+      auto& overwrite = *(++to_write.begin());
+      assert(!overwrite.is_existing());
+      assert(front.addr == front.pin->get_key());
+      ops.to_remap.push_back(extent_to_remap_t::create_overwrite(
+        std::move(front.pin),
+        overwrite.addr - front.pin->get_key(),
+        overwrite.len,
+        overwrite.to_write));
+    } else {
+      assert(to_write.size() == 2);
+      auto& overwrite = *(to_write.begin());
+      assert(back.addr > back.pin->get_key());
+      ops.to_remap.push_back(extent_to_remap_t::create_overwrite(
+        std::move(back.pin),
+        overwrite.addr - back.pin->get_key(),
+        overwrite.len,
+        overwrite.to_write));
+    }
+    ops.to_remove.pop_front();
+    logger().debug(
+      "ovewrite"
+      " to_remap list size: {}"
+      " to_insert list size: {}"
+      " to_remove list size: {}",
+      ops.to_remap.size(), ops.to_insert.size(), ops.to_remove.size());
+    return ops;
+  }
+
+  //prepare to_insert
+  for (auto &region : to_write) {
+    if (region.is_data()) {
+      assert(region.to_write.has_value());
+      ops.to_insert.push_back(extent_to_insert_t::create_data(
+        region.addr, region.len, region.to_write));
+    } else if (region.is_zero()) {
+      assert(!(region.to_write.has_value()));
+      ops.to_insert.push_back(extent_to_insert_t::create_zero(
+        region.addr, region.len));
+    }
+  }
+
+  //prepare to_remap, overwrite happens in multiple extents
+  if (front.is_existing()) {
+    assert(to_write.size() > 1);
+    assert(front.addr == front.pin->get_key());
+    ops.to_remap.push_back(extent_to_remap_t::create_remap(
+      std::move(front.pin),
+      0,
+      front.len));
+    ops.to_remove.pop_front();
+  }
+  if (back.is_existing()) {
+    assert(to_write.size() > 1);
+    assert(back.addr > back.pin->get_key());
```

Comment:
the right boundary should match in this case:
assert(back.addr + back.len == back.pin->get_key() + back.pin->get_length());